compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -27,26 +27,148 @@ pub enum AccessLevel {
     Public,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, Debug, HashStable, Default)]
+pub struct EffectiveVisibility {
+    public: Option<Visibility>,
+    exported: Option<Visibility>,
+    reachable: Option<Visibility>,
+    reachable_from_impl_trait: Option<Visibility>,
+}
+
+impl EffectiveVisibility {
+    fn get_field(&self, tag: AccessLevel) -> &Option<Visibility> {
+        match tag {
+            AccessLevel::Public => &self.public,
+            AccessLevel::Exported => &self.exported,
+            AccessLevel::Reachable => &self.reachable,
+            AccessLevel::ReachableFromImplTrait => &self.reachable_from_impl_trait,
+        }
+    }
+
+    fn get_mut_field(&mut self, tag: AccessLevel) -> &mut Option<Visibility> {
+        match tag {
+            AccessLevel::Public => &mut self.public,
+            AccessLevel::Exported => &mut self.exported,
+            AccessLevel::Reachable => &mut self.reachable,
+            AccessLevel::ReachableFromImplTrait => &mut self.reachable_from_impl_trait,
+        }
+    }
+
+    pub fn get(&self, tag: AccessLevel) -> Option<&Visibility> {
+        (*self.get_field(tag)).as_ref()
+    }
+
+    fn get_mut(&mut self, tag: AccessLevel) -> Option<&mut Visibility> {
+        (*self.get_mut_field(tag)).as_mut()
```

Comment:
```suggestion
        self.get_mut_field(tag).as_mut()
```
The parentheses here and in `fn get` shouldn't be necessary?

Reply:
Also I wouldn't bother creating these `fn get/get_mut/set` as separate functions and just inline them and use `get_(mut_)field` instead.

In the immutable case `.as_ref()` can be done immediately inside `fn get_field` if that's more convenient.